README
NAME
    Alien::VideoLAN::LibVLC - building, finding and using VLC binaries

VERSION
    Version 0.0_5

SYNOPSIS
    Alien::VideoLAN::LibVLC tries (in given order) during its installation:

    * Locate an already installed VLC via pkg-config.
    * Download prebuilt VLC binaries (if available for your platform). Not
    implemented.
    * Build VLC binaries from source codes (if possible on your system). Not
    implemented.

    Later you can use Alien::VideoLAN::LibVLC in your module that needs to
    link agains LibVLC and/or related libraries like this:

            # Sample Makefile.pl
            use ExtUtils::MakeMaker;
            use Alien::VideoLAN::LibVLC;

            WriteMakefile(
              NAME         => 'Any::VLC::Module',
              VERSION_FROM => 'lib/Any/VLC/Module.pm',
              LIBS         => Alien::VideoLAN::LibVLC->config('libs', [-lAdd_Lib]),
              INC          => Alien::VideoLAN::LibVLC->config('cflags'),
              # + additional params
            );

DESCRIPTION
    Please see Alien for the manifesto of the Alien namespace.

    In short `Alien::LibVLC' can be used to detect and get configuration
    settings from an installed VLC and related libraries. Based on your
    platform it (doesn't currently) offers the possibility to download and
    install prebuilt binaries or to build VLC & co. from source codes.

    The important facts:

    * The module does not modify in any way the already existing VLC
    installation on your system.
    * If you reinstall VLC libs on your system you do not need to reinstall
    Alien::VideoLAN::LibVLC (providing that you use the same directory for
    the new installation).
    * The prebuild binaries and/or binaries built from sources are always
    installed into perl module's 'share' directory.
    * If you use prebuild binaries and/or binaries built from sources it
    happens that some of the dynamic libraries (*.so, *.dll) will not
    automaticly loadable as they will be stored somewhere under perl
    module's 'share' directory. To handle this scenario
    Alien::VideoLAN::LibVLC offers some special functionality (see below).

METHODS
  config()
    This function is the main public interface to this module. These
    functions return string:

            Alien::VideoLAN::LibVLC->config('prefix');
            Alien::VideoLAN::LibVLC->config('version');
            Alien::VideoLAN::LibVLC->config('includedir');

    These functions return lists of strings:

            Alien::VideoLAN::LibVLC->config('ldflags');
            Alien::VideoLAN::LibVLC->config('cflags');

    On top of that this function supports special parameters:

            Alien::VideoLAN::LibVLC->config('ld_shared_libs');

    Returns a reference to a list of full paths to shared libraries (*.so,
    *.dll) that will be required for running the resulting binaries you have
    linked with VLC libs.

            Alien::VideoLAN::LibVLC->config('ld_paths');

    Returns a reference to a list of full paths to directories with shared
    libraries (*.so, *.dll) that will be required for running the resulting
    binaries you have linked with VLC libs.

    NOTE: config('ld_<something>') return an empty list/hash if you have
    decided to use VLC libraries already installed on your system. This
    concerns pkg-config usage.

  check_header()
    This function checks the availability of given header(s) when using
    compiler options provided by
    "Alien::VideoLAN::LibVLC->config('cflags')".

            Alien::VideoLAN::LibVLC->check_header('vlc.h');
            Alien::VideoLAN::LibVLC->check_header('vlc.h', 'libvlc_media.h');

    Returns 1 if all given headers are available, 0 otherwise.

  get_header_version()
    Tries to find a header file specified as a param in VLC prefix direcotry
    and based on "#define" macros inside this header file tries to get a
    version triplet. Only `libvlc_version.h' makes sense.

            Alien::VideoLAN::LibVLC->get_header_version('libvlc_version.h');

    Returns string like '1.2.3' or undef if not able to find and parse
    version info.

  `find_libvlc' *(deprecated)*
        Alien::VideoLAN::LibVLC->find_libvlc();
        Alien::VideoLAN::LibVLC->find_libvlc(version => '>= 1.1.9');
        Alien::VideoLAN::LibVLC->find_libvlc(version => '= 1.1.10',
                                             suppress_error_message => 1);

    Finds installed libvlc.

    If `version' parameter is specified, required version is needed. Check
    documentation of `pkg-config' for format of version.

    If `suppress_error_message' parameter is specified and is true, nothing
    will be put to STDERR if libvlc is not found.

    Returns hash with following fields:

    * version
        a string with version.

    * cflags
        arrayref of strings, e.g. `['-I/foo/bar']'

    * ldflags
        arrayref of strings, e.g. `['-L/foo/baz', '-lvlc']'

    If libvlc of specified version isn't found, croaks.

BUGS
    Please post issues and bugs at
    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-VideoLAN-LibVLC

ACKNOWLEDGEMENTS
            Thanks to authors of other Alien modules.
            Since 0.05 this module is based on L<Alien::SDL>.

COPYRIGHT
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Alien-VideoLAN-LibVLC

0.0_5   2011-08-12, 22:13 UTC+7
      - Use Alien::SDL as base for this module.

0.04    2011-06-24, 08:21 UTC+7
      - Don't use // operator, it doesn't work with perl 5.8

0.03    2011-06-23, 18:53 UTC+7
      - Add suppress_error_message parameter.
      - Add few links to See Also chapter.

0.02    2011-06-23, 09:17 UTC+7
      - Remove stub EXPORTS section from documentation.

0.01    2011-06-22, 22:01 UTC+7
      - First version, released on an unsuspecting world.



Build.PL
use strict;
use warnings;

use lib "inc";
use File::Spec::Functions qw(catdir catfile);
use My::Utility qw(check_config_script check_pkgconfig check_prebuilt_binaries check_src_build);

#### we need the platform-specific module
my %platforms =(
	# Unix = default, thus not listing all UNIX like systems
	MSWin32 => 'Windows',
);
my $package = 'My::Builder::' . ($platforms{$^O} || 'Unix');
print "Gonna use '$package' class ...\n";
eval "require $package" or die "Require '$package' failed: $@\n";

my $sdl_config;

#### Stadard Module::Builder stuff
my $build = $package->new(
	module_name          => 'Alien::VideoLAN::LibVLC',
	all_from             => 'lib/Alien/VideoLAN/LibVLC.pm',
	dist_abstract        => 'Use VLC libraries',
#	dist_abstract        => 'Get, Build and Use VLC libraries',
	dist_author          => 'Alexey Sokolov <alexey@alexeysokolov.co.cc>',
	license              => 'perl',
	requires             => {
		'File::Spec'       => '0',
		'File::Temp'       => '0',
		'File::ShareDir'   => '0',
		'File::Which'      => '0',
		'ExtUtils::CBuilder' => '0',
		'Capture::Tiny'    => '0.11',
		'ExtUtils::PkgConfig' => '0',
		'Carp'             => '0',
	},
	build_requires       => {  #need to have for running: ./Build (install|test)
		'File::Spec'       => '0',
		'File::Temp'       => '0',
		'File::ShareDir'   => '0',
		'ExtUtils::CBuilder' => '0',
		'File::Path'       => '2.08',
		'File::Fetch'      => '0.24',
		'File::Find'       => '0',
		'File::Which'      => '0',
		'Digest::SHA'      => '0',
		'Archive::Extract' => '0',
		'Archive::Tar'     => '0',
		'Archive::Zip'     => '0',
		'Module::Build'    => '0.36',
		'Text::Patch'      => '1.4',
	},
	configure_requires   => {  #need to have for running: perl Buil.PL
		'File::Spec'       => '0',
		'File::Path'       => '2.08',
		'File::Fetch'      => '0.24',
		'File::Find'       => '0',
		'File::Which'      => '0',
		'Digest::SHA'      => '0',
		'Archive::Extract' => '0',
		'Module::Build'    => '0.36',
		'Text::Patch'      => '1.4',
		'File::ShareDir'   => '0',
		'ExtUtils::PkgConfig' => '0',
		'Capture::Tiny'    => '0.11',
	},
#	meta_merge => {
#		resources  => {
#			bugtracker => 'http://sdlperl.ath.cx/projects/SDLPerl/query?component=Alien%3A%3ASDL&status=!closed',
#			repository => 'http://github.com/PerlGameDev/Alien-SDL'
#		}
#	},
#	get_options => { 'with-sdl-config' => { qw(type :s  store) => \$sdl_config } },
	dynamic_config => 1,
	create_readme => 1,
	share_dir => 'sharedir',
# sharedir is used for storing compiled/prebuilt binaries of SDL lib + related libraries
# avoid using 'share' name as M::B doe not handle well paths like /xx/yy/share/zz/ww/share/xx
);

my $choice;

# Search for possible methods, then prompt:
print "\nWelcome to Alien::VideoLAN::LibVLC module installation";
print "\n------------------------------------------------------\n\n";

#### check what options we have for our platform
my $rv; my @candidates = ();

if($rv=check_pkgconfig("libvlc >= 1.2")) {
	push @candidates, $rv;
};

if($rv=check_prebuilt_binaries($build->os_type)) {
	push @candidates, @{$rv};
};

if($build->can_build_binaries_from_sources() && ($rv=check_src_build($build->os_type))) {
	push @candidates, @{$rv};
};

push @candidates, { title => 'Quit installation' };

#### ask user what way to go
my $i                     = 1;
my $prompt_string         = "\nYou have the following options:\n";
my $recommended_candidate = 1;
foreach my $c (@candidates) {
	$recommended_candidate = $i if $c->{title} =~ /RECOMMENDED/;
	$prompt_string        .= "[" . $i++ . "] " . $c->{title} . "\n"
}
$prompt_string .= "\nWhat way do you wanna go?";
my $ans = $build->prompt($prompt_string, $recommended_candidate);

if(($ans>0) && ($ans<scalar(@candidates))) {
	$choice = $candidates[$ans-1];
}

#### store build params into 'notes'
if($choice) {
	print "Using \l$choice->{title}\n";
	$build->notes('build_params', $choice);
	$build->create_build_script();

	#### clean build_done stamp; force rebuild when running 'Build'
	$build->clean_build_done_marker;
}
else {
	$build->notes('build_params', undef); # just to be sure
	exit(0); # we want no reports from CPAN Testers in this case
}


inc/My/Utility.pm
package My::Utility;
use strict;
use warnings;
use base qw(Exporter);

our @EXPORT_OK = qw(check_config_script check_pkgconfig check_prebuilt_binaries check_prereqs_libs check_prereqs_tools check_src_build find_SDL_dir find_file check_header sed_inplace get_dlext);
use Config;
use ExtUtils::CBuilder;
use File::Spec::Functions qw(splitdir catdir splitpath catpath rel2abs);
use File::Find qw(find);
use File::Which;
use File::Copy qw(cp);
use Cwd qw(realpath);
use Capture::Tiny qw(capture);
use ExtUtils::PkgConfig;

our $cc = $Config{cc};
#### packs with prebuilt binaries
# - all regexps has to match: arch_re ~ $Config{archname}, cc_re ~ $Config{cc}, os_re ~ $^O
# - the order matters, we offer binaries to user in the same order (1st = preffered)
my $prebuilt_binaries = [
];

#### tarballs with source codes
my $source_packs = [
## the first set for source code build will be a default option
];

sub check_config_script
{
	my $script = shift || 'sdl-config';
	print "Gonna check config script...\n";
	print "(scriptname=$script)\n";
	my $devnull = File::Spec->devnull();
	my $version = `$script --version 2>$devnull`;
	return if($? >> 8);
	my $prefix = `$script --prefix 2>$devnull`;
	return if($? >> 8);
	$version =~ s/[\r\n]*$//;
	$prefix =~ s/[\r\n]*$//;
	#returning HASHREF
	return {
		title     => "Already installed SDL ver=$version path=$prefix",
		buildtype => 'use_config_script',
		script    => $script,
		prefix    => $prefix,
	};
}

sub check_pkgconfig {
	my $pkgname = shift;
	print "Gonna check pkg-config...\n";
	my %p;
	eval {
		capture {
			%p = ExtUtils::PkgConfig->find($pkgname);
		};
	};
	return if $@;
	my $prefix = ExtUtils::PkgConfig->variable($pkgname, 'prefix');
	return {
		title     => "Already installed VLC ver=$p{modversion} path=$prefix",
		buildtype => 'use_pkgconfig',
		pkg_config_path => $ENV{PKG_CONFIG_PATH},
	};
}

sub check_prebuilt_binaries
{
	print "Gonna check availability of prebuilt binaries ...\n";
	print "(os=$^O cc=$cc archname=$Config{archname})\n";
	my @good = ();
	foreach my $b (@{$prebuilt_binaries}) {
		if ( ($^O =~ $b->{os_re}) &&
			($Config{archname} =~ $b->{arch_re}) &&
			($cc =~ $b->{cc_re}) ) {
			$b->{buildtype} = 'use_prebuilt_binaries';
			push @good, $b;
		}
	}
	#returning ARRAY of HASHREFs (sometimes more than one value)
	return \@good;
}

sub check_src_build
{
	print "Gonna check possibility for building from sources ...\n";
	print "(os=$^O cc=$cc)\n";
	my @good = ();
	foreach my $p (@{$source_packs}) {
		$p->{buildtype} = 'build_from_sources';
		print "CHECKING prereqs for:\n\t$p->{title}\n";
		push @good, $p if check_prereqs($p);
	}
	return \@good;
}

sub check_prereqs_libs {
	my @libs = @_;
	my $ret  = 1;

	foreach my $lib (@libs) {
		my $found_lib          = '';
		my $found_inc          = '';
		my $inc_lib_candidates = {
			'/usr/local/include' => '/usr/local/lib',
			'/usr/include'       => '/usr/lib',
			'/usr/X11R6/include' => '/usr/X11R6/lib',
			'/usr/pkg/include'   => '/usr/pkg/lib',
		};

		if( -e '/usr/lib64'  && $Config{'myarchname'} =~ /64/) {
			$inc_lib_candidates->{'/usr/include'} = '/usr/lib64'
		}

		if( exists $ENV{SDL_LIB} && exists $ENV{SDL_INC} ) {
			$inc_lib_candidates->{$ENV{SDL_INC}} = $ENV{SDL_LIB};
		}

		my $header_map         = {
			'z'    => 'zlib',
			'jpeg' => 'jpeglib',
		};
		my $header             = (defined $header_map->{$lib}) ? $header_map->{$lib} : $lib;

		my $dlext = get_dlext();
		foreach (keys %$inc_lib_candidates) {
			my $ld = $inc_lib_candidates->{$_};
			next unless -d $_ && -d $ld;
			($found_lib) = find_file($ld, qr/[\/\\]lib\Q$lib\E[\-\d\.]*\.($dlext[\d\.]*|a|dll.a)$/);
			($found_inc) = find_file($_,  qr/[\/\\]\Q$header\E[\-\d\.]*\.h$/);
			last if $found_lib && $found_inc;
		}

		if($found_lib && $found_inc) {
			$ret &= 1;
		}
		else {
			my $reason = 'no-h+no-lib';
			$reason = 'no-lib' if !$found_lib && $found_inc;
			$reason = 'no-h' if $found_lib && !$found_inc;
			print "WARNING: required lib(-dev) '$lib' not found, disabling affected option ($reason)\n";
			$ret = 0;
		}
	}

	return $ret;
}

sub check_prereqs {
	my $bp  = shift;
	my $ret = 1;

	$ret &= check_prereqs_libs(@{$bp->{prereqs}->{libs}}) if defined $bp->{prereqs}->{libs};

	return $ret;
}

sub check_prereqs_tools {
	my @tools = @_;
	my $ret  = 1;

	foreach my $tool (@tools) {

		if((File::Which::which($tool) && -x File::Which::which($tool))
			|| ('pkg-config' eq $tool && defined $ENV{PKG_CONFIG} && $ENV{PKG_CONFIG}
				&& File::Which::which($ENV{PKG_CONFIG})
				&& -x File::Which::which($ENV{PKG_CONFIG}))) {
			$ret &= 1;
		}
		else {
			print "WARNING: required '$tool' not found\n";
			$ret = 0;
		}
	}

	return $ret;
}

sub find_file {
	my ($dir, $re) = @_;
	my @files;
	$re ||= qr/.*/;
	{
		#hide warning "Can't opendir(...): Permission denied - fix for http://rt.cpan.org/Public/Bug/Display.html?id=57232
		no warnings;
		find({ wanted => sub { push @files, rel2abs($_) if /$re/ }, follow => 1, no_chdir => 1 , follow_skip => 2}, $dir);
	};
	return @files;
}

sub find_SDL_dir {
	my $root = shift;
	my ($version, $prefix, $incdir, $libdir);
	return unless $root;

	# try to find SDL_version.h
	my ($found) = find_file($root, qr/SDL_version\.h$/i ); # take just the first one
	return unless $found;

	# get version info
	open(DAT, $found) || return;
	my @raw=<DAT>;
	close(DAT);
	my ($v_maj) = grep(/^#define[ \t]+SDL_MAJOR_VERSION[ \t]+[0-9]+/, @raw);
	$v_maj =~ s/^#define[ \t]+SDL_MAJOR_VERSION[ \t]+([0-9]+)[.\r\n]*$/$1/;
	my ($v_min) = grep(/^#define[ \t]+SDL_MINOR_VERSION[ \t]+[0-9]+/, @raw);
	$v_min =~ s/^#define[ \t]+SDL_MINOR_VERSION[ \t]+([0-9]+)[.\r\n]*$/$1/;
	my ($v_pat) = grep(/^#define[ \t]+SDL_PATCHLEVEL[ \t]+[0-9]+/, @raw);
	$v_pat =~ s/^#define[ \t]+SDL_PATCHLEVEL[ \t]+([0-9]+)[.\r\n]*$/$1/;
	return if (($v_maj eq '')||($v_min eq '')||($v_pat eq ''));
	$version = "$v_maj.$v_min.$v_pat";

	# get prefix dir
	my ($v, $d, $f) = splitpath($found);
	my @pp = reverse splitdir($d);
	shift(@pp) if(defined($pp[0]) && $pp[0] eq '');
	shift(@pp) if(defined($pp[0]) && $pp[0] eq 'SDL');
	if(defined($pp[0]) && $pp[0] eq 'include') {
		shift(@pp);
		@pp = reverse @pp;
		return (
			$version,
			catpath($v, catdir(@pp), ''),
			catpath($v, catdir(@pp, 'include'), ''),
			catpath($v, catdir(@pp, 'lib'), ''),
		);
	}
}

sub check_header {
	my ($cflags, @header) = @_;
	print STDERR "Testing header(s): " . join(', ', @header) . "\n";
	my $cb = ExtUtils::CBuilder->new(quiet => 1);
	my ($fs, $src) = File::Temp->tempfile('XXXXaa', SUFFIX => '.c', UNLINK => 1);
	my $inc = '';
	$inc .= "#include <$_>\n" for @header;  
	syswrite($fs, <<MARKER); # write test source code
#if defined(_WIN32) && !defined(__CYGWIN__)
#include <stdio.h>
/* GL/gl.h on Win32 requires windows.h being included before */
#include <windows.h>
#endif
$inc
int demofunc(void) { return 0; }

MARKER
close($fs);
my $obj = eval { $cb->compile( source => $src, extra_compiler_flags => $cflags); };
if($obj) {
	unlink $obj;
	return 1;
}
else {
	print STDERR "###TEST FAILED### for: " . join(', ', @header) . "\n";
	return 0;
}
}

sub sed_inplace {
	# we expect to be called like this:
	# sed_inplace("filename.txt", 's/0x([0-9]*)/n=$1/g');
	my ($file, $re) = @_;
	if (-e $file) {
		cp($file, "$file.bak") or die "###ERROR### cp: $!";
		open INPF, "<", "$file.bak" or die "###ERROR### open<: $!";
		open OUTF, ">", $file or die "###ERROR### open>: $!";
		binmode OUTF; # we do not want Windows newlines
		while (<INPF>) {
			eval( "$re" );
			print OUTF $_;
		}
		close INPF;
		close OUTF;
	}
}

sub get_dlext {
	if($^O =~ /darwin/) { # there can be .dylib's on a mac even if $Config{dlext} is 'bundle'
		return 'so|dylib|bundle';
	}
	elsif( $^O =~ /cygwin/)
	{  
		return 'la';
	}
	else {
		return $Config{dlext};
	}
}

1;


lib/Alien/VideoLAN/LibVLC.pm
package Alien::VideoLAN::LibVLC;
use strict;
use warnings;
use Alien::VideoLAN::LibVLC::ConfigData;
use File::ShareDir qw(dist_dir);
use File::Spec;
use File::Find;
use File::Spec::Functions qw(catdir catfile rel2abs);
use File::Temp;
use Capture::Tiny;
use Config;
use Carp;
use ExtUtils::PkgConfig;

=head1 NAME

Alien::VideoLAN::LibVLC - building, finding and using VLC binaries

=head1 VERSION

Version 0.0_5

=cut

our $VERSION = '0.0_5';
$VERSION = eval $VERSION;

=head1 SYNOPSIS

Alien::VideoLAN::LibVLC tries (in given order) during its installation:

=over

=item * Locate an already installed VLC via pkg-config.

=item * Download prebuilt VLC binaries (if available for your platform). Not implemented.

=item * Build VLC binaries from source codes (if possible on your system). Not implemented.

=back

Later you can use Alien::VideoLAN::LibVLC in your module that needs to link agains LibVLC
and/or related libraries like this:

	# Sample Makefile.pl
	use ExtUtils::MakeMaker;
	use Alien::VideoLAN::LibVLC;

	WriteMakefile(
	  NAME         => 'Any::VLC::Module',
	  VERSION_FROM => 'lib/Any/VLC/Module.pm',
	  LIBS         => Alien::VideoLAN::LibVLC->config('libs', [-lAdd_Lib]),
	  INC          => Alien::VideoLAN::LibVLC->config('cflags'),
	  # + additional params
	);

=head1 DESCRIPTION

Please see L<Alien> for the manifesto of the Alien namespace.

In short C<Alien::LibVLC> can be used to detect and get
configuration settings from an installed VLC and related libraries.
Based on your platform it (doesn't currently) offers the possibility to download and
install prebuilt binaries or to build VLC & co. from source codes.

The important facts:

=over

=item * The module does not modify in any way the already existing VLC
installation on your system.

=item * If you reinstall VLC libs on your system you do not need to
reinstall Alien::VideoLAN::LibVLC (providing that you use the same directory for
the new installation).

=item * The prebuild binaries and/or binaries built from sources are always
installed into perl module's 'share' directory.

=item * If you use prebuild binaries and/or binaries built from sources
it happens that some of the dynamic libraries (*.so, *.dll) will not
automaticly loadable as they will be stored somewhere under perl module's
'share' directory. To handle this scenario Alien::VideoLAN::LibVLC offers some special
functionality (see below).

=back

=head1 METHODS

=head2 config()

This function is the main public interface to this module. These functions
return string:

	Alien::VideoLAN::LibVLC->config('prefix');
	Alien::VideoLAN::LibVLC->config('version');
	Alien::VideoLAN::LibVLC->config('includedir');

These functions return lists of strings:

	Alien::VideoLAN::LibVLC->config('ldflags');
	Alien::VideoLAN::LibVLC->config('cflags');

On top of that this function supports special parameters:

	Alien::VideoLAN::LibVLC->config('ld_shared_libs');

Returns a reference to a list of full paths to shared libraries (*.so, *.dll) that will be
required for running the resulting binaries you have linked with VLC libs.

	Alien::VideoLAN::LibVLC->config('ld_paths');

Returns a reference to a list of full paths to directories with shared libraries (*.so, *.dll)
that will be required for running the resulting binaries you have linked with
VLC libs.

NOTE: config('ld_<something>') return an empty list/hash if you have decided to
use VLC libraries already installed on your system. This concerns pkg-config usage.

=head2 check_header()

This function checks the availability of given header(s) when using compiler
options provided by "Alien::VideoLAN::LibVLC->config('cflags')".

	Alien::VideoLAN::LibVLC->check_header('vlc.h');
	Alien::VideoLAN::LibVLC->check_header('vlc.h', 'libvlc_media.h');

Returns 1 if all given headers are available, 0 otherwise.

=head2 get_header_version()

Tries to find a header file specified as a param in VLC prefix direcotry and
based on "#define" macros inside this header file tries to get a version triplet.
Only C<libvlc_version.h> makes sense.

	Alien::VideoLAN::LibVLC->get_header_version('libvlc_version.h');

Returns string like '1.2.3' or undef if not able to find and parse version info.

=head2 C<find_libvlc> I<(deprecated)>

    Alien::VideoLAN::LibVLC->find_libvlc();
    Alien::VideoLAN::LibVLC->find_libvlc(version => '>= 1.1.9');
    Alien::VideoLAN::LibVLC->find_libvlc(version => '= 1.1.10',
                                         suppress_error_message => 1);

Finds installed libvlc.

If C<version> parameter is specified, required version is needed.
Check documentation of C<pkg-config> for format of version.

If C<suppress_error_message> parameter is specified and is true,
nothing will be put to STDERR if libvlc is not found.

Returns hash with following fields:

=over 4

=item * B<version>

a string with version.

=item * B<cflags>

arrayref of strings, e.g. C<['-I/foo/bar']>

=item * B<ldflags>

arrayref of strings, e.g. C<['-L/foo/baz', '-lvlc']>

=back

If libvlc of specified version isn't found, croaks.

=head1 BUGS

Please post issues and bugs at L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-VideoLAN-LibVLC>

=head1 ACKNOWLEDGEMENTS

	Thanks to authors of other Alien modules.
	Since 0.05 this module is based on L<Alien::SDL>.

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut

### get config params
sub config
{
	my $package = shift;
	my @params  = @_;
	my $bt = Alien::VideoLAN::LibVLC::ConfigData->config('build_params')->{buildtype};
	return &_libvlc_config_via_pkgconfig if $bt eq 'use_pkgconfig';
	return _libvlc_config_via_script(@params)      if(Alien::VideoLAN::LibVLC::ConfigData->config('script'));# there is not such script...
	return _libvlc_config_via_config_data(@params) if(Alien::VideoLAN::LibVLC::ConfigData->config('config'));# need to install vlc by ours, to use it.
}

### get version info from given header file
sub get_header_version {
	my ($package, $header) = @_;
	return unless $header;

	# try to find header
	my $include = Alien::VideoLAN::LibVLC->config('includedir');
	my @files;
	find({ wanted => sub { push @files, rel2abs($_) if /\Q$header\E$/ }, follow => 1, no_chdir => 1, follow_skip => 2 }, $include);
	return unless @files;

	# get version info
	open(DAT, $files[0]) || return;
	my @raw=<DAT>;
	close(DAT);
	print for @raw;

	# generic magic how to get version major/minor/patchlevel
	my ($v_maj) = grep(/^#[ \t]*define[ \t]+[A-Z_]+?MAJOR[A-Z_]*[ \t]+\([0-9]+\)/, @raw);
	$v_maj =~ s/^#[ \t]*define[ \t]+[A-Z_]+[ \t]+\(([0-9]+)\)[.\r\n]*$/$1/;
	my ($v_min) = grep(/^#[ \t]*define[ \t]+[A-Z_]+MINOR[A-Z_]*[ \t]+\([0-9]+\)/, @raw);
	$v_min =~ s/^#[ \t]*define[ \t]+[A-Z_]+[ \t]+\(([0-9]+)\)[.\r\n]*$/$1/;
	my ($v_pat) = grep(/^#[ \t]*define[ \t]+[A-Z_]+REVISION[A-Z_]*[ \t]+\([0-9]+\)/, @raw);
	$v_pat =~ s/^#[ \t]*define[ \t]+[A-Z_]+[ \t]+\(([0-9]+)\)[.\r\n]*$/$1/;
	my ($v_ext) = grep(/^#[ \t]*define[ \t]+[A-Z_]+EXTRA[A-Z_]*[ \t]+\([0-9]+\)/, @raw);
	$v_ext =~ s/^#[ \t]*define[ \t]+[A-Z_]+[ \t]+\(([0-9]+)\)[.\r\n]*$/$1/;
	return if (($v_maj eq '')||($v_min eq '')||($v_pat eq '')||($v_ext eq ''));
	my $ver = "$v_maj.$v_min.$v_pat";
	$ver .= ".$v_ext" if $v_ext ne '0';
	return $ver;
}

### check presence of header(s) specified as params
sub check_header {
	my ($package, @header) = @_;
	print STDERR "[$package] Testing header(s): " . join(', ', @header);

	require ExtUtils::CBuilder; # PAR packer workaround

	my $config  = {};
	if($^O eq 'cygwin') {
		my $ccflags = $Config{ccflags};
		$ccflags    =~ s/-fstack-protector//;
		$config     = { ld => 'gcc', cc => 'gcc', ccflags => $ccflags };
	}

	my $cb = ExtUtils::CBuilder->new( quiet => 1, config => $config );
	my ($fs, $src) = File::Temp->tempfile('XXXXaa', SUFFIX => '.c', UNLINK => 1);
	my $inc = '';
	$inc .= "#include <$_>\n" for @header;
	syswrite($fs, <<MARKER); # write test source code
#if defined(_WIN32) && !defined(__CYGWIN__)
/* GL/gl.h on Win32 requires windows.h being included before */
#include <windows.h>
#endif
$inc
int demofunc(void) { return 0; }

MARKER
	close($fs);
	my $obj;
	my $stdout = '';
	my $stderr = '';
	($stdout, $stderr) = Capture::Tiny::capture {
		$obj = eval { $cb->compile( source => $src, extra_compiler_flags => Alien::VideoLAN::LibVLC->config('cflags')); };
	};
	if($obj) {
		print STDERR "\n";
		unlink $obj;
		return 1;
	}
	else {
		$stderr =~ s/[\r\n]$//;
		$stderr =~ s/^\Q$src\E[\d\s:]*//;

		print STDERR " NOK: ($stderr)\n";
		return 0;
	}
}

### internal functions
sub _libvlc_config_via_pkgconfig {
	my $PKGNAME = 'libvlc >= 1.2';
	my $param = shift;
	my $path = Alien::VideoLAN::LibVLC::ConfigData->config('build_params')->{pkg_config_path};
	$path = "" unless $path;
	$path = "$path:$ENV{PKG_CONFIG_PATH}" if $ENV{PKG_CONFIG_PATH};
	local $ENV{PKG_CONFIG_PATH} = $path;
	return ExtUtils::PkgConfig->modversion($PKGNAME) if $param eq 'version';
	return grep { $_ ne '' } split /\s/, ExtUtils::PkgConfig->cflags($PKGNAME) if $param eq 'cflags';
	return grep { $_ ne '' } split /\s/, ExtUtils::PkgConfig->libs($PKGNAME) if $param eq 'ldflags';
	return [] if $param =~ /^ld_/;
	return ExtUtils::PkgConfig->variable($PKGNAME, $param);
}

sub _libvlc_config_via_script
{
	croak "there's no libvlc-config script around";
	my $param    = shift;
	my @add_libs = @_;
	my $devnull = File::Spec->devnull();
	my $script = Alien::SDL::ConfigData->config('script');
	return unless ($script && ($param =~ /[a-z0-9_]*/i));
	my $val = `$script --$param 2>$devnull`;
	$val =~ s/[\r\n]*$//;
	if($param eq 'cflags') {
		$val .= ' ' . Alien::SDL::ConfigData->config('additional_cflags');
	}
	elsif($param eq 'libs') {
		$val .= ' ' . join(' ', @add_libs) if scalar @add_libs;
		$val .= ' ' . Alien::SDL::ConfigData->config('additional_libs');
	}
	return $val;
}

sub _libvlc_config_via_config_data
{
	croak "vlc installation by ourselves isn't supported yet";
	my $param    = shift;
	my @add_libs = @_;
	my $share_dir = dist_dir('Alien-VideoLAN-LibVLC');
	my $subdir = Alien::VideoLAN::LibVLC::ConfigData->config('share_subdir');
	return unless $subdir;
	my $real_prefix = catdir($share_dir, $subdir);
	return unless ($param =~ /[a-z0-9_]*/i);
	my $val = Alien::VideoLAN::LibVLC::ConfigData->config('config')->{$param};
	return unless $val;
	# handle additional flags
	if($param eq 'cflags') {
		$val .= ' ' . Alien::VideoLAN::LibVLC::ConfigData->config('additional_cflags');
	}
	elsif($param eq 'libs') {
		$val .= ' ' . join(' ', @add_libs) if scalar @add_libs;
		$val .= ' ' . Alien::VideoLAN::LibVLC::ConfigData->config('additional_libs');
	}  
	# handle @PrEfIx@ replacement
	if ($param =~ /^(ld_shared_libs|ld_paths)$/) {
		s/\@PrEfIx\@/$real_prefix/g foreach (@{$val});
	}
	elsif ($param =~ /^(ld_shlib_map)$/) {
		while (my ($k, $v) = each %$val ) {
			$val->{$k} =~ s/\@PrEfIx\@/$real_prefix/g;
		}
	}
	else {
		$val =~ s/\@PrEfIx\@/$real_prefix/g;
	}
	return $val;
}

### deprecated stuff
sub _find {
	my $self = shift;
	my $lib = shift;
	my %a = @_;

	my $version = $a{version};
	$version = '' unless defined $version;
	my %p;

	if ($a{suppress_error_message}) {
		my $str;
		open my $fh, '>', \$str;
		local *STDERR = $fh;
		%p = ExtUtils::PkgConfig->find("$lib $version");
	} else {
		%p = ExtUtils::PkgConfig->find("$lib $version");
	}

	my @cflags = grep { $_ ne '' } split /\s/, $p{cflags};
	$p{cflags} = \@cflags;
	my @ldflags = grep { $_ ne '' } split /\s/, $p{libs};
	delete $p{libs};
	$p{ldflags} = \@ldflags;
	$p{version} = $p{modversion};
	delete $p{modversion};
	return %p;
}
sub find_libvlc {
	my $self = shift;
	my %a = @_;
	return $self->_find('libvlc', %a);
}


1;
